kv: crash on dropping database

I was trying to run some queries from the photo app locally and I noticed a crash.

This reproduces it:
```
DROP DATABASE TEST;
CREATE DATABASE TEST;
SET DATABASE=TEST;

CREATE TABLE IF NOT EXISTS users (
  id           INT,
  photoCount   INT,
  commentCount INT,
  name         STRING,
  address      STRING,
  PRIMARY KEY (id)
);
CREATE TABLE IF NOT EXISTS photos (
  id           BYTES DEFAULT EXPERIMENTAL_UUID_V4(),
  userID       INT,
  commentCount INT,
  caption      STRING,
  latitude     FLOAT,
  longitude    FLOAT,
  timestamp    TIMESTAMP,
  PRIMARY KEY (id)
);
CREATE UNIQUE INDEX IF NOT EXISTS photosByUserID ON photos (userID, timestamp);
CREATE TABLE IF NOT EXISTS comments (
  photoID   BYTES,
  commentID BYTES DEFAULT EXPERIMENTAL_UUID_V4(),
  userID    INT,
  message   STRING,
  timestamp TIMESTAMP,
  PRIMARY KEY (photoID, commentID)
);
CREATE UNIQUE INDEX IF NOT EXISTS commentsByPhotoID ON comments (photoID, timestamp);

DROP DATABASE TEST;
```

```
panic: attempt to clone resolve_intent_range:<header:<key:"\305" end_key:"\307" > intent_txn:<id:<1437:2699763392 9:/* unknown wire type 6 */ 4:/* unknown wire type 6 */ 28478418:/* unexpected EOF */ /* unexpected EOF */ >isolation:SERIALIZABLE key:"\307" epoch:0 timestamp:<wall_time:1457986634287956693 logical:0 > > status:COMMITTED poison:false > , which contains ed59c01e-eba0-4e26-90bd-d16c8f9eae96

goroutine 1085 [running]:
panic(0x13e5200, 0xc8207877e0)
	/usr/local/go/src/runtime/panic.go:464 +0x3e6
github.com/cockroachdb/cockroach/util.CloneProto(0x7f8c57d0df00, 0xc820113148, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/util/clone_proto.go:60 +0x26d
github.com/cockroachdb/cockroach/kv.truncate(0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc8205ea520, 0xc82057e000, ...)
	/go/src/github.com/cockroachdb/cockroach/kv/batch.go:99 +0x70c
github.com/cockroachdb/cockroach/kv.(*DistSender).sendChunk.func1(0xc8206c4da0, 0xc8206c4af0, 0xc8206c50d0, 0xc82030e980, 0x7f8c57d1f3a8, 0xc8205e2750, 0x8, 0xc820847d70)
	/go/src/github.com/cockroachdb/cockroach/kv/dist_sender.go:607 +0x11d
github.com/cockroachdb/cockroach/kv.(*DistSender).sendChunk(0xc82030e980, 0x7f8c57d1f4b0, 0xc8205b0f30, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/kv/dist_sender.go:619 +0x16cf
github.com/cockroachdb/cockroach/kv.(*DistSender).Send(0xc82030e980, 0x7f8c57d1f4b0, 0xc8205b0f30, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/kv/dist_sender.go:490 +0x42c
github.com/cockroachdb/cockroach/kv.(*TxnCoordSender).Send(0xc82041e2a0, 0x7f8c57d1f4b0, 0xc8205b0f30, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/kv/txn_coord_sender.go:411 +0xce3
github.com/cockroachdb/cockroach/client.(*DB).send(0xc82030d040, 0xc8202b6800, 0x12, 0x20, 0x0, 0xc8205ea4e0)
	/go/src/github.com/cockroachdb/cockroach/client/db.go:413 +0x16c
github.com/cockroachdb/cockroach/client.(*DB).(github.com/cockroachdb/cockroach/client.send)-fm(0xc8202b6800, 0x12, 0x20, 0x1, 0x7f8c57d994b0)
	/go/src/github.com/cockroachdb/cockroach/client/db.go:378 +0x3e
github.com/cockroachdb/cockroach/client.sendAndFill(0xc8206c5888, 0xc8207b6380, 0x0, 0x10)
	/go/src/github.com/cockroachdb/cockroach/client/db.go:343 +0x4d
github.com/cockroachdb/cockroach/client.(*DB).RunWithResponse(0xc82030d040, 0xc8207b6380, 0xc8209c1980, 0x5d7bb2)
	/go/src/github.com/cockroachdb/cockroach/client/db.go:378 +0x6b
github.com/cockroachdb/cockroach/client.(*DB).Run(0xc82030d040, 0xc8207b6380, 0xc8205ea4e0)
	/go/src/github.com/cockroachdb/cockroach/client/db.go:369 +0x2b
github.com/cockroachdb/cockroach/storage.(*Replica).resolveIntents.func3(0x1632060)
	/go/src/github.com/cockroachdb/cockroach/storage/replica.go:1977 +0x31
github.com/cockroachdb/cockroach/storage.(*Replica).resolveIntents(0xc82069e0e0, 0x7f8c57d1f618, 0xc82005a300, 0xc8205cb000, 0x12, 0x22, 0x1, 0x0)
	/go/src/github.com/cockroachdb/cockroach/storage/replica.go:1986 +0x9f5
github.com/cockroachdb/cockroach/storage.(*Replica).handleSkippedIntents.func1()
	/go/src/github.com/cockroachdb/cockroach/storage/replica.go:1808 +0x2ec
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunAsyncTask.func1(0xc820301180, 0x1f2027d, 0x12, 0x740, 0xc82072f590)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:140 +0x62
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunAsyncTask
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:141 +0xc9
```

It hits every time, but strangely doesn't hit if I comment out one of the tables and its index..

No cockroach-data directory. The store is reported as `store[0]:  type=mem,size=954 MiB`.

Tried restarting without the --store arg and verified it created a fresh directory, same crash.

Something could be getting corrupted.. Here are some more messages from reproducing:

```
panic: attempt to clone resolve_intent_range:<header:<key:"\273" end_key:"\275" > intent_txn:<id:<389050636076:/* unexpected EOF */ 9:9226937799293643451 >isolation:SERIALIZABLE key:"\275" epoch:0 timestamp:<wall_time:1457996311636402072 logical:0 > > status:COMMITTED poison:false > , which contains e2f2b2ce-ca5a-4b49-bbaa-85c30aab0c80

panic: attempt to clone resolve_intent_range:<header:<key:"\273" end_key:"\275" > intent_txn:<id:<872 { 673:/* unexpected EOF */ 8:/* unexpected EOF */ 177392:/* unknown wire type 7 */ 901:/* unexpected EOF */ >isolation:SERIALIZABLE key:"\275" epoch:0 timestamp:<wall_time:1457996469689671052 logical:0 > > status:COMMITTED poison:false > , which contains c3368a2a-ca55-42d8-a5b9-0087cf56ad38

panic: attempt to clone resolve_intent_range:<header:<key:"\273" end_key:"\275" > intent_txn:<id:<227633:/* unexpected EOF */ 2:2543209542 67 { 5 { 0 { 10:/* unknown wire type 7 */ /* unexpected EOF */ >isolation:SERIALIZABLE key:"\275" epoch:0 timestamp:<wall_time:1457996486854650181 logical:0 > > status:COMMITTED poison:false > , which contains 8a936f4e-1546-4c96-979b-042b035791f6

```

--- a/kv/batch.go
+++ b/kv/batch.go
@@ -93,14 +93,14 @@ func truncate(ba roachpb.BatchRequest, rs roachpb.RSpan) (roachpb.BatchRequest, 
 			ba.Requests[pos] = union
 		} else {
 			// Keep the old one. If we must adjust the header, must copy.
-			// TODO(tschottdorf): this could wind up cloning big chunks of data.
-			// Can optimize by creating a new Request manually, but with the old
-			// data.
-			if newHeader.Equal(origRequests[pos].GetInner().Header()) {
+			if inner := origRequests[pos].GetInner(); newHeader.Equal(inner.Header()) {
 				ba.Requests[pos] = origRequests[pos]
 			} else {
-				ba.Requests[pos] = *util.CloneProto(&origRequests[pos]).(*roachpb.RequestUnion)
-				ba.Requests[pos].GetInner().SetHeader(newHeader)
+				shallowCopy := inner.ShallowCopy()
+				shallowCopy.SetHeader(newHeader)
+				if union := &ba.Requests[pos]; !union.SetInner(shallowCopy) {
+					panic(fmt.Sprintf("%T excludes %T", union, shallowCopy))
+				}
 			}
 		}
 		if err != nil {

--- a/kv/truncate_test.go
+++ b/kv/truncate_test.go
@@ -8,8 +8,8 @@ import (
 	"github.com/cockroachdb/cockroach/keys"
 	"github.com/cockroachdb/cockroach/roachpb"
 	"github.com/cockroachdb/cockroach/testutils"
-	"github.com/cockroachdb/cockroach/util"
 	"github.com/cockroachdb/cockroach/util/leaktest"
+	"github.com/cockroachdb/cockroach/util/uuid"
 )
 
 func TestTruncate(t *testing.T) {
@@ -92,8 +92,9 @@ func TestTruncate(t *testing.T) {
 		goldenOriginal := roachpb.BatchRequest{}
 		for _, ks := range test.keys {
 			if len(ks[1]) > 0 {
-				goldenOriginal.Add(&roachpb.ScanRequest{
-					Span: roachpb.Span{Key: roachpb.Key(ks[0]), EndKey: roachpb.Key(ks[1])},
+				goldenOriginal.Add(&roachpb.ResolveIntentRangeRequest{
+					Span:      roachpb.Span{Key: roachpb.Key(ks[0]), EndKey: roachpb.Key(ks[1])},
+					IntentTxn: roachpb.TxnMeta{ID: uuid.NewV4()},
 				})
 			} else {
 				goldenOriginal.Add(&roachpb.GetRequest{
@@ -102,7 +103,10 @@ func TestTruncate(t *testing.T) {
 			}
 		}
 
-		original := *util.CloneProto(&goldenOriginal).(*roachpb.BatchRequest)
+		original := roachpb.BatchRequest{Requests: make([]roachpb.RequestUnion, len(goldenOriginal.Requests))}
+		for i, request := range goldenOriginal.Requests {
+			original.Requests[i].SetValue(request.GetInner().ShallowCopy())
+		}
 
 		desc := &roachpb.RangeDescriptor{
 			StartKey: roachpb.RKey(test.desc[0]), EndKey: roachpb.RKey(test.desc[1]),

--- a/roachpb/api.go
+++ b/roachpb/api.go
@@ -109,6 +109,8 @@ type Request interface {
 	SetHeader(Span)
 	// Method returns the request method.
 	Method() Method
+	// ShallowCopy returns a shallow copy of the receiver.
+	ShallowCopy() Request
 	createReply() Response
 	flags() int
 }
@@ -309,13 +311,13 @@ func (sr *ScanRequest) SetBound(bound int64) {
 }
 
 // GetBound returns the MaxResults field in ReverseScanRequest.
-func (sr *ReverseScanRequest) GetBound() int64 {
-	return sr.MaxResults
+func (rsr *ReverseScanRequest) GetBound() int64 {
+	return rsr.MaxResults
 }
 
 // SetBound sets the MaxResults field in ReverseScanRequest.
-func (sr *ReverseScanRequest) SetBound(bound int64) {
-	sr.MaxResults = bound
+func (rsr *ReverseScanRequest) SetBound(bound int64) {
+	rsr.MaxResults = bound
 }
 
 // Countable is implemented by response types which have a number of
@@ -408,6 +410,156 @@ func (*ComputeChecksumRequest) Method() Method { return ComputeChecksum }
 
 // Method implements the Request interface.
 func (*VerifyChecksumRequest) Method() Method { return VerifyChecksum }
+
+// ShallowCopy implements the Request interface.
+func (gr *GetRequest) ShallowCopy() Request {
+	shallowCopy := *gr
+	return &shallowCopy
+}
+
+// ShallowCopy implements the Request interface.
+func (pr *PutRequest) ShallowCopy() Request {
+	shallowCopy := *pr
+	return &shallowCopy
+}
+
+// ShallowCopy implements the Request interface.
+func (cpr *ConditionalPutRequest) ShallowCopy() Request {
+	shallowCopy := *cpr
+	return &shallowCopy
+}
+
+// ShallowCopy implements the Request interface.
+func (ir *IncrementRequest) ShallowCopy() Request {
+	shallowCopy := *ir
+	return &shallowCopy
+}
+
+// ShallowCopy implements the Request interface.
+func (dr *DeleteRequest) ShallowCopy() Request {
+	shallowCopy := *dr
+	return &shallowCopy
+}
+
+// ShallowCopy implements the Request interface.
+func (drr *DeleteRangeRequest) ShallowCopy() Request {
+	shallowCopy := *drr
+	return &shallowCopy
+}
+
+// ShallowCopy implements the Request interface.
+func (sr *ScanRequest) ShallowCopy() Request {
+	shallowCopy := *sr
+	return &shallowCopy
+}
+
+// ShallowCopy implements the Request interface.
+func (rsr *ReverseScanRequest) ShallowCopy() Request {
+	shallowCopy := *rsr
+	return &shallowCopy
+}
+
+// ShallowCopy implements the Request interface.
+func (ccr *CheckConsistencyRequest) ShallowCopy() Request {
+	shallowCopy := *ccr
+	return &shallowCopy
+}
+
+// ShallowCopy implements the Request interface.
+func (btr *BeginTransactionRequest) ShallowCopy() Request {
+	shallowCopy := *btr
+	return &shallowCopy
+}
+
+// ShallowCopy implements the Request interface.
+func (etr *EndTransactionRequest) ShallowCopy() Request {
+	shallowCopy := *etr
+	return &shallowCopy
+}
+
+// ShallowCopy implements the Request interface.
+func (asr *AdminSplitRequest) ShallowCopy() Request {
+	shallowCopy := *asr
+	return &shallowCopy
+}
+
+// ShallowCopy implements the Request interface.
+func (amr *AdminMergeRequest) ShallowCopy() Request {
+	shallowCopy := *amr
+	return &shallowCopy
+}
+
+// ShallowCopy implements the Request interface.
+func (htr *HeartbeatTxnRequest) ShallowCopy() Request {
+	shallowCopy := *htr
+	return &shallowCopy
+}
+
+// ShallowCopy implements the Request interface.
+func (gcr *GCRequest) ShallowCopy() Request {
+	shallowCopy := *gcr
+	return &shallowCopy
+}
+
+// ShallowCopy implements the Request interface.
+func (ptr *PushTxnRequest) ShallowCopy() Request {
+	shallowCopy := *ptr
+	return &shallowCopy
+}
+
+// ShallowCopy implements the Request interface.
+func (rlr *RangeLookupRequest) ShallowCopy() Request {
+	shallowCopy := *rlr
+	return &shallowCopy
+}
+
+// ShallowCopy implements the Request interface.
+func (rir *ResolveIntentRequest) ShallowCopy() Request {
+	shallowCopy := *rir
+	return &shallowCopy
+}
+
+// ShallowCopy implements the Request interface.
+func (rirr *ResolveIntentRangeRequest) ShallowCopy() Request {
+	shallowCopy := *rirr
+	return &shallowCopy
+}
+
+// ShallowCopy implements the Request interface.
+func (nr *NoopRequest) ShallowCopy() Request {
+	shallowCopy := *nr
+	return &shallowCopy
+}
+
+// ShallowCopy implements the Request interface.
+func (mr *MergeRequest) ShallowCopy() Request {
+	shallowCopy := *mr
+	return &shallowCopy
+}
+
+// ShallowCopy implements the Request interface.
+func (tlr *TruncateLogRequest) ShallowCopy() Request {
+	shallowCopy := *tlr
+	return &shallowCopy
+}
+
+// ShallowCopy implements the Request interface.
+func (llr *LeaderLeaseRequest) ShallowCopy() Request {
+	shallowCopy := *llr
+	return &shallowCopy
+}
+
+// ShallowCopy implements the Request interface.
+func (ccr *ComputeChecksumRequest) ShallowCopy() Request {
+	shallowCopy := *ccr
+	return &shallowCopy
+}
+
+// ShallowCopy implements the Request interface.
+func (vcr *VerifyChecksumRequest) ShallowCopy() Request {
+	shallowCopy := *vcr
+	return &shallowCopy
+}
 
 func (*GetRequest) createReply() Response                { return &GetResponse{} }
 func (*PutRequest) createReply() Response                { return &PutResponse{} }

--- a/storage/replica_test.go
+++ b/storage/replica_test.go
@@ -3333,11 +3333,11 @@ func testRangeDanglingMetaIntent(t *testing.T, isReverse bool) {
 	var origSeen, newSeen bool
 
 	for !(origSeen && newSeen) {
-		clonedRLArgs := util.CloneProto(rlArgs).(*roachpb.RangeLookupRequest)
+		clonedRLArgs := *rlArgs
 
 		reply, pErr = client.SendWrappedWith(tc.Sender(), tc.rng.context(), roachpb.Header{
 			ReadConsistency: roachpb.INCONSISTENT,
-		}, clonedRLArgs)
+		}, &clonedRLArgs)
 		if pErr != nil {
 			t.Fatal(pErr)
 		}
@@ -3428,11 +3428,11 @@ func TestRangeLookupUseReverseScan(t *testing.T) {
 
 	// Test ReverseScan without intents.
 	for _, c := range testCases {
-		clonedRLArgs := util.CloneProto(rlArgs).(*roachpb.RangeLookupRequest)
+		clonedRLArgs := *rlArgs
 		clonedRLArgs.Key = keys.RangeMetaKey(roachpb.RKey(c.key))
 		reply, pErr := client.SendWrappedWith(tc.Sender(), tc.rng.context(), roachpb.Header{
 			ReadConsistency: roachpb.INCONSISTENT,
-		}, clonedRLArgs)
+		}, &clonedRLArgs)
 		if pErr != nil {
 			t.Fatal(pErr)
 		}
@@ -3457,11 +3457,11 @@ func TestRangeLookupUseReverseScan(t *testing.T) {
 
 	// Test ReverseScan with intents.
 	for _, c := range testCases {
-		clonedRLArgs := util.CloneProto(rlArgs).(*roachpb.RangeLookupRequest)
+		clonedRLArgs := *rlArgs
 		clonedRLArgs.Key = keys.RangeMetaKey(roachpb.RKey(c.key))
 		reply, pErr := client.SendWrappedWith(tc.Sender(), tc.rng.context(), roachpb.Header{
 			ReadConsistency: roachpb.INCONSISTENT,
-		}, clonedRLArgs)
+		}, &clonedRLArgs)
 		if pErr != nil {
 			t.Fatal(pErr)
 		}

